WI #619: `SKU Price Details` preferred keys (#683)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: specification/columns/skupricedetails.md

Reviewer: Since we currently have 3 PRs addressing the SkuPriceDetails Normative Requirements, I’m posting a preview so you can get an idea of how it might look if all three are approved.

You will notice that I’ve included a few suggestions on how to modify the 3 normative requirements we just added in this PR.

Additionally, I suggest renaming **Predefined keys** to **FOCUS-defined SKU Price properties** (or something similar), as in the table, we have not only defined the key for a property but also outlined additional requirements.

Note: This is just a suggestion, and I’d really appreciate any feedback, @kk09v , especially from you! 😄 

---

The SkuPriceDetails column adheres to the following requirements:

* SkuPriceDetails MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset) when the provider supports unit pricing concepts and publishes [*price lists*](#glossary:price-list), publicly or as part of contracting.
* SkuPriceDetails MUST be of type String.
* SkuPriceDetails MUST conform to [StringHandling](#stringhandling) requirements.
* SkuPriceDetails MUST conform to [KeyValueFormat](#key-valueformat) requirements.
* Property key SHOULD conform to [PascalCase](#glossary:pascalcase) format.
* SkuPriceDetails nullability is defined as follows:
  * SkuPriceDetails MUST be null when SkuPriceId is null.
  * SkuPriceDetails MAY be null when SkuPriceId is not null.
* When SkuPriceDetails is not null, SkuPriceDetails adheres to the following additional requirements:
  * SkuPriceDetails MUST be associated with a given [ResourceId](#resourceid) or [ServiceName](#servicename) used or purchased.
  * SkuPriceDetails MUST NOT include properties that are not applicable to the corresponding SkuPriceId.
  * SkuPriceDetails SHOULD include all FOCUS-defined SKU Price properties listed below that are applicable to the corresponding SkuPriceId.
  * SkuPriceDetails MAY include properties that are already captured in other dedicated columns.
  * SkuPriceDetails SHOULD remain consistent over time for a given SkuPriceId.
  * SkuPriceDetails properties for a given SkuPriceId adhere to the following additional requirements:
    * Existing SkuPriceDetails properties SHOULD remain consistent over time.
    * Existing SkuPriceDetails properties SHOULD NOT be removed.
    * Additional SkuPriceDetails properties MAY be added over time.
  * Property key SHOULD remain consistent across comparable SKUs having that property, and the values for this key SHOULD remain in a consistent format.
  * Property value MUST represent the value for a single [PricingUnit](#pricingunit) when the property holds a numeric value.
* FOCUS-defined SKU Price properties adhere to the following additional requirements:
  * Property key MUST equal the predefined key for that property.
  * Property value MUST be of the type specified for that property.
  * Property value MUST represent the value for a single unit of measure specified for that property when the property holds a numeric value.

Author: To make this easier to compare, I took the existing statements and put them in the same order/indentation to compare the differences and evaluate whether this changed the meaning. Any lines below in ALL CAPS are my summary of a newly added line.

Notes (beginning with line number of your suggested rephrasing, including the opening line and following blank line)
*  7 I'd change this to keys instead of key
* 12 What is intended here? SkuPriceDetails should not be associated with a resource ID. Why would we specify an association to Service Name instead of to SkuId / SkuPriceId? Sorry, I'm just not following the reasoning on this in my first read.
* 14 this is a change in meaning from what was written (addition for 1.2), but I think I'm okay with it as it mostly acts to strengthen the language from "if you include it, you should..." to "you should include it"
`SkuPriceDetails properties that represent common constructs listed in the FOCUS-Defined Properties section below SHOULD use the defined keys.`
`SkuPriceDetails SHOULD include all FOCUS-defined SKU Price properties listed below that are applicable to the corresponding SkuPriceId.`
* 16 seems redundant with lines 17-20
* 26 "the value for a single unit of measure" didn't quite make sense to me. I'd suggest the below. The for a single PricingUnit clause above (22) already handles an aspect of this, though I've included it again since it's a SHOULD on 22.
`Property value MUST represent the value for a single PricingUnit, denominated in the unit of measure specified for that property when the property holds a numeric value.`
* + I added a new clause related to the FOCUS-defined property namespacing between your lines 22 and 23
* - I noticed that a clause was dropped in your suggestion. This is somewhat implied, but I think it's important because it makes it clear that you shouldn't include for example a resource property in here (which would differ across different records for the same SkuPriceId).
`SkuPriceDetails properties (both keys and values) MUST be shared across all charges having the same SkuPriceId, subject to the below provisions.`

---
* SkuPriceDetails MUST be present in a FOCUS dataset when the provider includes a SkuPriceId.
* STRING TYPE
* STRING Handling
* SkuPriceDetails MUST be in KeyValueFormat.
* SkuPriceDetails keys SHOULD be formatted in PascalCase.
* NULL grouping
  * SkuPriceDetails MAY be null when SkuPriceId is not null.
  * SkuPriceDetails MUST be null when SkuPriceId is null.
* PROPERTY GROUPING
  * ASSOCIATED…
  * SkuPriceDetails MUST NOT contain properties which are not applicable to the corresponding SkuPriceId.
  * SkuPriceDetails properties that represent common constructs listed in the FOCUS-Defined Properties section below SHOULD use the defined keys.
  * SkuPriceDetails MAY contain properties which are already captured in other dedicated columns.
  * SKU CONSISTENCY
    * Individual properties SHOULD NOT be modified for a given SkuPriceId and SHOULD remain consistent over time.
    * Properties SHOULD NOT be removed from SkuPriceDetails for a given SkuPriceId, once they have been included.
    * Additional properties MAY be added to SkuPriceDetails going forward for a given SkuPriceId.
  * The key for a property SHOULD remain consistent across comparable SKUs having that property and the values for this key SHOULD remain in a consistent format.
  * SkuPriceDetails properties with a numeric value MUST represent the value for a single PricingUnit.
  * SkuPriceDetails keys MUST NOT begin with the string "F_" unless it is a FOCUS-defined property
* FOCUS PROPERTIES
  * PROPERTY KEYS MUST MATCH
  * Values for FOCUS-defined properties MUST be the specified data type for that property in the section below.
  * Numeric values for FOCUS-defined properties MUST be the specified unit of measure for that property in the section below.

Reviewer: Cleaned up version:

---

The SkuPriceDetails column adheres to the following requirements:

* SkuPriceDetails MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset) when the provider supports unit pricing concepts and publishes [*price lists*](#glossary:price-list), publicly or as part of contracting.
* SkuPriceDetails MUST be of type String.
* SkuPriceDetails MUST conform to [StringHandling](#stringhandling) requirements.
* SkuPriceDetails MUST conform to [KeyValueFormat](#key-valueformat) requirements.
* Property keys SHOULD conform to [PascalCase](#glossary:pascalcase) format.
* SkuPriceDetails nullability is defined as follows:
  * SkuPriceDetails MUST be null when SkuPriceId is null.
  * SkuPriceDetails MAY be null when SkuPriceId is not null.
* When SkuPriceDetails is not null, SkuPriceDetails adheres to the following additional requirements:
  * SkuPriceDetails MUST be associated with a given SkuPriceId.
  * SkuPriceDetails MUST NOT include properties that are not applicable to the corresponding SkuPriceId.
  * SkuPriceDetails SHOULD include all FOCUS-defined SKU Price properties listed below that are applicable to the corresponding SkuPriceId.
  * SkuPriceDetails MAY include properties that are already captured in other dedicated columns.
  * SkuPriceDetails properties for a given SkuPriceId adhere to the following additional requirements:
    * Existing SkuPriceDetails properties SHOULD remain consistent over time.
    * Existing SkuPriceDetails properties SHOULD NOT be removed.
    * Additional SkuPriceDetails properties MAY be added over time.
  * Property key SHOULD remain consistent across comparable SKUs having that property, and the values for this key SHOULD remain in a consistent format.
  * Property value MUST represent the value for a single [PricingUnit](#pricingunit) when the property holds a numeric value.
* FOCUS-defined SKU Price properties adhere to the following additional requirements:
  * Property key MUST equal the predefined key for that property.
  * Property value MUST be of the type specified for that property.
  * Property value MUST represent the value for a single PricingUnit, denominated in the unit of measure specified for that property when the property holds a numeric value.

Author: Thanks again @ijurica , I just reviewed this before merging and it looks like the F_ prefix clause got missed. I know you were going to look for where to fit that bullet in, but I'll put it next to the other "FOCUS-defined SKU Price Properties clause" for now.

Reviewer: @kk09v I wonder if we need two separate requirements - one specifying when the F_ (or x_ 😄 ) prefix MUST be included and another specifying when it MUST NOT be included.

Author: Because we're prefixing FOCUS properties and we've said that they must not prefix provider properties with F_, we have one side covered (must not)
By stating
`Property key MUST equal the predefined key for that property.`
and defining the key for the properties to have a F_ prefix, we have the the other side covered (must)

The SkuPriceDetails column adheres to the following requirements:

* SkuPriceDetails MUST be in [KeyValueFormat](#key-valueformat).
* SkuPriceDetails keys SHOULD be formatted in [PascalCase](#glossary:pascalcase).
* SkuPriceDetails properties (both keys and values) MUST be shared across all charges having the same SkuPriceId, subject to the below provisions.
  * Additional properties MAY be added to SkuPriceDetails going forward for a given SkuPriceId.
  * Properties SHOULD NOT be removed from SkuPriceDetails for a given SkuPriceId, once they have been included.
  * Individual properties SHOULD NOT be modified for a given SkuPriceId and SHOULD remain consistent over time.

> * SkuPriceDetails keys and their values SHOULD remain consistent across comparable SKUs having that property.

Reviewer: Phrasing aside, I prefer the original version of this normative requirement because I have some concerns regarding the material change introduced by this revision. While the new version is suitable for SKU-related properties, it is not suitable for Pricing properties.

Maybe we should split this in two requirements - one for SKU-related properties and the other one for Pricing?

```suggestion
* The key for a property SHOULD remain consistent across comparable SKUs having that property and the values for this key SHOULD remain in a consistent format.
```

Author: Because the diff got hosed during multiple rewrites. I want to clarify this this suggestion is a revert to the prior phrasing from 1.1  

I agree that this attempt at clearer phrasing substantially changed the meaning of the clause. Thank you for catching this, I had missed it when I read through the proposed changes the first time.

The intention of this clause was that if a provider offers two "comparable" SKUs (e.g. two different virtual machines), we would not want one virtual machine to have a property of {"CoreCount":8} and the other to have a property of {"vCPUs":"eight"}. Ignoring everything else that's wrong with the second example, the keys should be the same and the values should have the same format.

Author: For now, I'll commit this suggestion so that the meaning is not changed, but if we can have another go at rephrasing this for clarity because the change in meaning due to rephase is a solid indication that the original phrasing is not clear.

The SkuPriceDetails column adheres to the following requirements:

* SkuPriceDetails MUST be in [KeyValueFormat](#key-valueformat).
* SkuPriceDetails keys SHOULD be formatted in [PascalCase](#glossary:pascalcase).
* SkuPriceDetails properties (both keys and values) MUST be shared across all charges having the same SkuPriceId, subject to the below provisions.
  * Additional properties MAY be added to SkuPriceDetails going forward for a given SkuPriceId.
  * Properties SHOULD NOT be removed from SkuPriceDetails for a given SkuPriceId, once they have been included.
  * Individual properties SHOULD NOT be modified for a given SkuPriceId and SHOULD remain consistent over time.
* The key for a property SHOULD remain consistent across comparable SKUs having that property and the values for this key SHOULD remain in a consistent format.
* SkuPriceDetails MUST NOT contain properties which are not applicable to the corresponding SkuPriceId.
* SkuPriceDetails MAY contain properties which are already captured in other dedicated columns.
* SkuPriceDetails properties with a [numeric](#numeric-format) value MUST represent the value for a single [PricingUnit](#pricingunit).

> * SkuPriceDetails keys MUST NOT begin with the string "F_" unless it is a FOCUS-defined property.

Reviewer 2: I remember when this came up on Friday and was holding my comments until we had a chance to discuss, but we ran out of time. I really, really, really don't like this. While I understand the concept, it's the opposite of what we do for custom columns. I would highly prefer something that aligns with the current pattern for custom attributes, if necessary at all.

Reviewer: As I mentioned in the meeting, I understand the reasoning behind how it's currently specified, but I’m more inclined to apply the same principle as for column naming.

Reviewer 3: I agree _strongly_ with @flanakin on this one: this is the opposite of what we do for custom columns.  We should have no prefix for FOCUS-defined properties, similar to how we handle FOCUS-defined columns.

Author: This has been taken to a vote in #working-group per the TF2 meeting today, so I'll close this conversation.

* SkuPriceDetails properties that represent common constructs listed in the FOCUS-Defined Properties section below SHOULD use the defined keys.
  * Values for FOCUS-defined properties MUST be the specified data type for that property in the section below.
  * Numeric values for FOCUS-defined properties MUST be the specified unit of measure for that property in the section below.
* SkuPriceDetails MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset) when the provider includes a SkuPriceId.

>   * SkuPriceDetails MAY be null when SkuPriceId is not null.

Reviewer 2: I've heard some people say that they want an empty property bag instead of null. Thoughts? @shawnalpay may have an opinion 😉

Reviewer: For what it's worth, I really, really prefer null over empty strings and bags 😄 

And here is one of the String Handling requirements:
> * Empty strings and strings consisting solely of spaces SHOULD NOT be used in not-nullable string columns.

Reviewer 3: Oh geez.  I agree with @ijurica on this one.  As a practitioner, I might interpret an empty property bag as a data error, and I'd rather get null.  Also aligns with the principle/requirement Irena quoted.

Reviewer: Even though it's old and dusty, I'm pulling my Java programmer hat out of the drawer ... and I still prefer `null`. 😄

  * SkuPriceDetails MUST be null when SkuPriceId is null.

## Examples

```json
{
    "AccessTier": "Archive",
    "CoreCount": 4,

>     "PreimumProcessing": true,

Reviewer 2: ```suggestion
    "PremiumProcessing": true,
```

Author: I had intentionally left this in there as an example of a theoretical provider-defined property with a Boolean value.

Reviewer 3: @kk09v Just to be clear: you intentionally left a typo...?  `Preimum` instead of `Premium`.

# SKU Price Details

SKU Price Details represent a list of [*SKU Price*](#glossary:sku-price) properties (key-value pairs) associated with a specific [SKU Price ID](#skupriceid). These properties include qualitative and quantitative properties of a [*SKUs*](#glossary:sku) (e.g., functionality and technical specifications), along with core stable pricing properties (e.g., pricing terms, tiers, etc.), excluding dynamic or negotiable pricing elements such as unit price amounts, currency (and related exchange rates), temporal validity (e.g., effective dates), and contract- or negotiation-specific factors (e.g., contract or account identifiers, and negotiable discounts).

The composition of properties associated with a specific *SKU Price* may differ across providers and across *SKUs* within the same provider. However, the exclusion of dynamic or negotiable pricing properties should ensure that all charges with the same SKU Price ID share the same SKU Price Details, i.e., that SKU Price Details remains consistent across different billing periods and billing accounts within a provider.

> SKU Price Details helps practitioners understand and distinguish *SKU Prices*, each identified by a SKU Price ID and associated with a used or purchased [*resource*](#glossary:resource) or [*service*](#glossary:service). It can also enable practitioners to calculate metrics such as total units of a SKU when it is not directly priced in those units (e.g. cores) and thus enables FinOps capabilities such as unit economics. Additionally, the SKU Price Details may be used to analyze costs based on pricing properties such as terms and tiers.

Reviewer: just a suggestion 😄 

```suggestion
SKU Price Details helps practitioners understand and distinguish *SKU Prices*, each identified by a SKU Price ID and associated with a used or purchased [*resource*](#glossary:resource) or [*service*](#glossary:service). It can also help determine the quantity of units for a property when it holds a numeric value (e.g., CoreCount), even when its unit differs from the one in which the SKU is priced and charged, thus supporting FinOps capabilities like unit economics. Additionally, the SKU Price Details may be used to analyze costs based on pricing properties such as terms and tiers.
```

SKU Price Details helps practitioners understand and distinguish *SKU Prices*, each identified by a SKU Price ID and associated with a used or purchased [*resource*](#glossary:resource) or [*service*](#glossary:service). It can also enable practitioners to calculate metrics such as total units of a SKU when it is not directly priced in those units (e.g. cores) and thus enables FinOps capabilities such as unit economics. Additionally, the SKU Price Details may be used to analyze costs based on pricing properties such as terms and tiers.

The SkuPriceDetails column adheres to the following requirements:

* SkuPriceDetails MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset) when the provider supports unit pricing concepts and publishes [*price lists*](#glossary:price-list), publicly or as part of contracting.

> * SkuPriceDetails MUST be of type String.
> * SkuPriceDetails MUST conform to [StringHandling](#stringhandling) requirements.

Reviewer: Per @UdamDewaraja suggestion and TF-1 discussion reverting back to JSON (as specified in FOCUS 1.1).

_**Note:** This will be revisited in future releases (AI #855)._ 

```suggestion
```

  * SkuPriceDetails MUST be associated with a given SkuPriceId.
  * SkuPriceDetails MUST NOT include properties that are not applicable to the corresponding SkuPriceId.
  * SkuPriceDetails SHOULD include all FOCUS-defined SKU Price properties listed below that are applicable to the corresponding SkuPriceId.
  * SkuPriceDetails MAY include properties that are already captured in other dedicated columns.
  * SkuPriceDetails properties for a given SkuPriceId adhere to the following additional requirements:
    * Existing SkuPriceDetails properties SHOULD remain consistent over time.
    * Existing SkuPriceDetails properties SHOULD NOT be removed.
    * Additional SkuPriceDetails properties MAY be added over time.
  * Property key SHOULD remain consistent across comparable SKUs having that property, and the values for this key SHOULD remain in a consistent format.
  * Property key MUST NOT begin with the string "F_" unless it is a FOCUS-defined property.
  * Property value MUST represent the value for a single [PricingUnit](#pricingunit) when the property holds a numeric value.
* FOCUS-defined SKU Price properties adhere to the following additional requirements:

>   * Property key MUST equal the predefined key for that property.

Reviewer 2: I think this is saying that, if you define a core count, you MUST use "CoreCount" and not "NumberOfCores" or "CpuCount" or something else.

>   * Property key MUST equal the predefined key for that property.

Reviewer 4: Given that this is normative, i suggest we have clear guidance here. If it seems unnecessary, maybe we remove it for now? Open to a different way of saying it if the goal is to emphasize what Michael said. 

```suggestion
```

>   * Property key MUST equal the predefined key for that property.

Reviewer 2: Doesn't removing this nullify the entire body of work because it then won't require data generators use the defined keys? I think we need to fix it if there's confusion. Does this help?
```suggestion
  * Property key MUST be one of the predefined key names when there is a logical equivalent for the information provided.
```

>   * Property key MUST equal the predefined key for that property.

Reviewer 3: We agreed in Mar 31 Maintainers that @kk09v is now armed with the crux of this suggestion (i.e. "use a FOCUS property key if it's applicable"), and he will wordsmith a more accurate wording.

>   * Property key MUST equal the predefined key for that property.

Author: I clarified this bullet as `Property key MUST match the spelling and casing specified for the FOCUS-defined property.`

>   * Property key MUST equal the predefined key for that property.

Author: There was an additional concern raised today around forcing inclusion of the FOCUS-defined property. We already had the provision indicating that all the FOCUS-defined properties SHOULD be included when applicable.
`SkuPriceDetails SHOULD include all FOCUS-defined SKU Price properties listed below that are applicable to the corresponding SkuPriceId.`
to which I added a subsequent bullet:
`SkuPriceDetails MUST include the FOCUS-defined SKU Price property when an equivalent property is included as a Provider-defined property.`

So, if a provider is already sending a property `x_vcpus` which represents the number of vCPUs on the SKU as a Provider-defined property, this would require them to also provide `CoreCount` once it's introduced as a FOCUS-defined property.

Technically, the subsequent bullet may be out of order per the normative requirements, but I think that it makes the most sense if this immediately follows the other one.

>   * Property key MUST equal the predefined key for that property.

Author: Udam reviewed and approved this PR which I'll interpret to reflect he's happy with the changes. I'm going to mark this conversation resolved.

The SkuPriceDetails column adheres to the following requirements:

* SkuPriceDetails MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset) when the provider supports unit pricing concepts and publishes [*price lists*](#glossary:price-list), publicly or as part of contracting.
* SkuPriceDetails MUST be of type String.
* SkuPriceDetails MUST conform to [StringHandling](#stringhandling) requirements.
* SkuPriceDetails MUST conform to [KeyValueFormat](#key-valueformat) requirements.

> * Property keys SHOULD conform to [PascalCase](#glossary:pascalcase) format.

Reviewer 2: nit: I would like each requirement to be able to stand on its own, when possible. Adding the column name to be explicit.
```suggestion
* SkuPriceDetails property keys SHOULD conform to [PascalCase](#glossary:pascalcase) format.
```

Reviewer: I've already applied that suggestion in the #798 😄 

@kk09v I’ve also slightly reformulated the Guidelines. (I’m trying to find what works best for multiple columns, and yes - there are exceptions 😄).

JSON columns are specific in that we specify requirements for:
- the column itself (Tags, SkuPriceDetails)
- the elements it is made up of (properties and tags)
- property keys/values and tag keys/values

We should probably specify that each requirement starts with the name of the subject (most often a column name) ?!

## Content Constraints

| Constraint    | Value                              |
| :------------ | :--------------------------------- |
| Column type   | Dimension                          |
| Feature level | Conditional                        |
| Allows nulls  | True                               |
| Data type     | String                             |
| Value format  | [KeyValueFormat](#key-valueformat) |

### FOCUS-Defined Properties

> The following keys should be used when applicable to facilitate cross-service and cross-provider queries for the same conceptual property. FOCUS-defined keys will be prefixed with "F_" to make them easy to identify and prevent collisions with other properties conveyed by a provider.

Reviewer: something to consider

```suggestion
The following keys should be used when applicable to facilitate cross-SKU and cross-provider queries for the same conceptual property. FOCUS-defined keys will be prefixed with "F_" to make them easy to identify and prevent collisions with other properties conveyed by a provider.
```

| Key                        | Description                                                     | Data Type        | Unit of Measure (numeric) or example values (string)  |
| :------------------------- | :-------------------------------------------------------------- | :--------------- | :---------------------------------------------------- |

> | F_Bandwidth                | Network capacity for data transfer                              | Numeric          | Measure: Megabits per second (Mbps)                   |

Reviewer 2: This "F_" is F_ing killing me 🤢😒 This makes FOCUS feel like a second-class citizen in its own spec. In a world where FOCUS properties are the majority, would we seriously have this across the board? That would be horrible. We should be prioritizing our space and requiring custom prefixes for custom properties. Same as with columns.

Reviewer 3: Strong agree.  An `F_` prefix feels wrong.  Must we have any prefix convention here at all?  Is it perhaps sufficient to just have a list of reserved FOCUS keys that thou shalt not use?

Author: Here's my understanding of Riley's thinking when he proposed we prefix the FOCUS columns:
- Providers will ideally be introducing properties that are relevant to their SKUs, independent of the FOCUS-defined properties.
- If a provider adds a property {"XylophoneUnits" : 12} and following year xylophone computing becomes the next big thing to the point that we want to add that as a FOCUS-defined property, then we've created a collision which is problematic because we've both said that for FOCUS-defined property keys they MUST give values in the format we've specified but also that they SHOULD NOT remove/change a property once it's been introduced.
  - By reserving a prefix, we never need to worry about these collisions.

The alternative would be to force a prefix onto any property keys that aren't FOCUS defined. 

I can go either way on which gets prefixed. However, what makes me nervous about this conversation is Michael's statement "In a world where FOCUS properties are the majority" implying that he is not intending to include many properties that aren't FOCUS-defined. I haven't done the math, but I suspect we have a long tail in terms of SKUs with 20% making up 80% of the spend. We've been focused on the 20% of the SKUs, particularly the ones that have commonalities across multiple CSPs for defining these properties. However, all of the PaaS offerings which are unique to each CSP should have their own properties, which would never become a FOCUS-defined property.

Reviewer 3: > The alternative would be to force a prefix onto any property keys that aren't FOCUS defined.

I feel like you did not address the question in my comment.  Can you tell me a little more about why we must force a prefix on _any_ property keys?  

> Must we have any prefix convention here at all? Is it perhaps sufficient to just have a list of reserved FOCUS keys that thou shalt not use?

Author: Sorry, I find examples and illustrations to be helpful for me to understand things, so I often explain that way. Let me try this again.

- Until a property key is part of the list of FOCUS-Defined properties, it could be adopted by a provider as the key for a provider-defined property.
  - To illustrate this, I started out with placeholder property "XyzProperty.", before making it more concrete by morphing it into "XylophoneUnits"
- If in a later release, FOCUS defines a property with a key that is already in use by a provider, this becomes problematic based on a few clauses:
  - Summarizing [3 clauses](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/blob/000b45b78f5e2643a11cd82ec1be61de32107095/specification/columns/skupricedetails.md?plain=1#L27) Property keys, value formats, and units of measure for FOCUS-defined properties MUST follow what is laid out for that FOCUS-defined property.
  - Now what happens if the FOCUS-defined property differs in description or unit of measure from what was contained in the provider-defined column?
    - E.g. FOCUS wants XylophoneUnits to mean the number of bars on the Xylophone, but the provider had defined it as the number of notes per octave of the scale that the Xylophone is tuned to (BTW, I'm **really** stretching for these examples since xylophone was just my replacement for xyz originally).
    - Properties SHOULD remain consistent over time and SHOULD NOT be removed [link](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/blob/000b45b78f5e2643a11cd82ec1be61de32107095/specification/columns/skupricedetails.md?plain=1#L21)
      - This is only a SHOULD, but depending on your interpretation, the provider would either be removing their property if the value specified in FOCUS is not something they have available or changing the value of the property to match what is defined by FOCUS.
    - The provider also now needs to define a new key for their existing property since they can no longer use the key reserved for FOCUS.
    - Also consider the timing of implementation. Providers are not necessarily adopting the newest version of FOCUS the moment it's released, so unless a practitioner is digging into the metadata to find which version of FOCUS their feed is compatible with and reviewing the changelog/prior versions of FOCUS to determine when the FOCUS-defined property was introduced, there's a strong chance of them misinterpreting the meaning of this property to be aligned to the FOCUS definition for that key.

Reviewer 3: I grant you that there may be a collision.  However, it presumes that the working group would not conduct a survey of keys in use by FOCUS providers, and I should hope that we would execute on a good-faith effort to avoid any existing key names.  If we do decide to enact [eminant domain](https://en.wikipedia.org/wiki/Eminent_domain), then I would hope that we do it for a very good reason.

I would very much like to avoid requiring practitioners to have to chop off a prefix during their FOCUS transformations process before getting the data into a data warehouse or semantic model for BI purposes.

I'm willing to stand down on this point if it's a non-starter with the group -- we'll see what the Slack poll generates -- but if so, then my fallback preference would be, as stated previously, to prefix the provider keys, as that has follows our precedent of requiring providers to prefix their non-FOCUS columns with `x_`.

Reviewer: Although I generally prefer reserved keywords, Karl pointed out my main concern:

> * _The provider also now needs to define a new key for their existing property since they can no longer use the key reserved for FOCUS.._

That's the main reason I opted for `x_`

| F_InstanceType             | Common name of the instance including size, shape, series, etc. | String           | Examples: "m5d.2xlarge", "NC24rs_v3", "P50"           |
| F_MemorySize               | RAM allocated for processing                                    | Numeric          | Measure: Gibibytes (GiB<sup>1</sup>)                  |
| F_OperatingSystem          | Operating system family<sup>2</sup>                             | String           | Examples: "Linux", "MacOS", "Windows"                 |
| F_Redundancy               | Level of redundancy offered by the SKU                          | String           | Examples: "Local", "Zonal", "Global"                  |
| F_Series                   | Common name for the series and/or generation of the SKU         | String           | Examples: "M5", "Dadv5", "N2D"                        |
| F_SizeClass                | Logical grouping of size, which may be used for size fexibility | String           | Examples: "Large", "2xlarge", "Metal"                 |

> | F_StorageClass             | Tier or class of storage provided                               | String           | Examples: "Hot", "Archive", "Nearline"                |

Reviewer: ```suggestion
| F_StorageClass             | Class or tier of storage provided                               | String           | Examples: "Hot", "Archive", "Nearline"                |
```

Reviewer 2: Should we have preferred values for things like this to ensure consistency when an applicable value exists?

Author: I had looked at that, and at first blush there are a lot of similarities.
Azure has Hot, Cool, Cold, Archive
GCP has Standard, Nearline, Coldline, Archive
AWS has Standard, Infrequent Access, and Archive

These sound similar, but the specific capabilities differ and we'd need to specify what constitutes each Class we establish (e.g. should Infrequent Access be grouped with Cool or Cold storage?).

Then there are the special things each offers: Azure Premium, AWS Express One Zone, and auto tiering options like S3 Intelligent Tiering and GCP Autoclass. GCP Autoclass still falls under the existing categories, but I can't speak to how AWS does it.

Finally, we don't have a set of preferred values for any preferred key yet, so we'd need to build a bit more around that. TBH, I'm not actually opposed to this idea, but I don't think we have the time to get to a consensus on that by the deadline given how I've understood the work required.

| :------------ | :--------------------------------- |
| Column type   | Dimension                          |
| Feature level | Conditional                        |
| Allows nulls  | True                               |
| Data type     | JSON                               |
| Value format  | [KeyValueFormat](#key-valueformat) |

### FOCUS-Defined Properties

The following keys should be used when applicable to facilitate cross-SKU and cross-provider queries for the same conceptual property. Focus-defined keys will appear in the list below and Provider-defined keys will be prefixed with "x_" to make them easy to identify as well as prevent collisions.

| Key                      | Description                                                     | Data Type        | Unit of Measure (numeric) or example values (string)  |
| :----------------------- | :-------------------------------------------------------------- | :--------------- | :---------------------------------------------------- |
| CoreCount                | Number of physical or virtual CPUs available<sup>1</sup>        | Numeric          | Measure: Quantity of Cores                            |

> | DiskIops                 | Storage input/output operations per second<sup>1</sup>          | Numeric          | Measure: Input/Output Operations per Second (IOPS)    |

Reviewer 3: Did the TF discuss use of the word `Max` in this key name?  Typically, IOPS capabilities are expressed as Max IOPS to delineate from any observability data, where IOPS is a metric and not an attribute.  Examples: [Azure](https://learn.microsoft.com/en-us/azure/virtual-machines/disks-types), [AWS](https://docs.aws.amazon.com/ebs/latest/userguide/ebs-volume-types.html), and [GCP](https://cloud.google.com/compute/docs/disks/performance).
```suggestion
| DiskMaxIops                 | Storage maximum input/output operations per second<sup>1</sup>          | Numeric          | Measure: Input/Output Operations per Second (IOPS)    |
```

| DiskSpace                | Storage capacity available                                      | Numeric          | Measure: Gibibytes (GiB)                              |
| DiskType                 | Kind of disk used                                               | String           | Examples: "SSD", "HDD", "NVMe"                        |
| GpuCount                 | Number of GPUs available                                        | Numeric          | Measure: Quantity of GPUs                             |
| InstanceType             | Common name of the instance including size, shape, series, etc. | String           | Examples: "m5d.2xlarge", "NC24rs_v3", "P50"           |
| InstanceSeries           | Common name for the series and/or generation of the instance    | String           | Examples: "M5", "Dadv5", "N2D"                        |
| MemorySize               | RAM allocated for processing                                    | Numeric          | Measure: Gibibytes (GiB<sup>2</sup>)                  |

> | NetworkBandwidth         | Network capacity for data transfer<br><sup>1</sup>              | Numeric          | Measure: Megabits per second (Mbps)                   |

Reviewer 3: Was there TF discussion of the use of `Bandwidth` vs `Throughput`?  The latter seems like the much more common publicly-facing description / pricing of network capabilities, and as a vendor, that was the term we used.

Examples:
https://learn.microsoft.com/en-us/azure/vpn-gateway/about-gateway-skus
https://docs.aws.amazon.com/storagegateway/latest/vgw/PerfGatewayAWS-common.html
https://cloud.google.com/compute/all-pricing?hl=en

Author: It was not discussed as nobody had taken issue with what was proposed. Michael had originally made the suggestion as simply "Bandwidth" which was later renamed NetworkBandwidth when another property was prefixed with Network.

Author: I can go either way on this, depending on what we're looking at, but I understand the argument for Throughput and am leaning more and more that direction.

To get everyone on the same starting page, Bandwidth measures capacity and Throughput measures usage. Here's an example explanation https://www.geeksforgeeks.org/difference-between-bandwidth-and-throughput/

Your [AWS example](https://docs.aws.amazon.com/storagegateway/latest/vgw/PerfGatewayAWS-common.html) is specifically talking about the measurement of actual bytes, not the capacity.
In one of your other [AWS examples](https://docs.aws.amazon.com/ebs/latest/userguide/ebs-volume-types.html), it talks about Max Throughput, so it's a choice between NetworkBandwidth and NetworkMaxThroughput. Those sound like the same thing, but I understand there's a subtle difference.
IEEE uses Bandwidth when talking about the ports and [cables](https://ieeexplore.ieee.org/document/STDGTSU20531) that are carrying the data. However, I understand that unless I'm buying a bare metal instance, I'm not getting a 10/100/1000 port, I'm getting an allocation of network capacity which is dependent on the provider's network stack and thus should be measured by observation of actual performance (aka NetworkMaxThroughput).

Author: After further pondering, I updated to NetworkMaxThroughput to bring it in line with DiskMaxIops 
@shawnalpay I'd like to make the same change NetworkIops -> NetworkMaxIops, but I wanted to gut check that one with your first.

Author: Resolving this conversation and moving that prior comment into it's own suggestion

> | NetworkIops              | Network input/output operations per second<sup>1</sup>                   | Numeric          | Measure: Input/Output Operations per Second (IOPS)    |

Author: ```suggestion
| NetworkMaxIops           | Network maximum sustained input/output operations per second<sup>1</sup> | Numeric          | Measure: Input/Output Operations per Second (IOPS)    |
```

Author: Suggestion to bring NetworkIops in line with DiskMaxIops and NetworkMaxThroughput.

Reviewer 3: Definitely.  +1.

## Content Constraints

| Constraint    | Value                              |
| :------------ | :--------------------------------- |
| Column type   | Dimension                          |
| Feature level | Conditional                        |
| Allows nulls  | True                               |

> | Data type     | String                             |

Reviewer: Per @UdamDewaraja suggestion and TF-1 discussion reverting back to JSON (as specified in FOCUS 1.1).

_**Note:** This will be revisited in future releases (AI #855)._ 

```suggestion
| Data type     | JSON                               |
```

| Value format  | [KeyValueFormat](#key-valueformat) |

### FOCUS-Defined Properties

The following keys should be used when applicable to facilitate cross-SKU and cross-provider queries for the same conceptual property. Focus-defined keys will appear in the list below and Provider-defined keys will be prefixed with "x_" to make them easy to identify as well as prevent collisions.

| Key                      | Description                                                     | Data Type        | Unit of Measure (numeric) or example values (string)  |
| :----------------------- | :-------------------------------------------------------------- | :--------------- | :---------------------------------------------------- |
| CoreCount                | Number of physical or virtual CPUs available<br><sup>1</sup>    | Numeric          | Measure: Quantity of Cores                            |

> | DiskInputOutput          | Storage input/output operations per second<br><sup>1</sup>      | Numeric          | Measure: Input/Output Operations per Second (IOPS)    |

Reviewer 4: ```suggestion
| DiskIOPS          | Storage input/output operations per second<br><sup>1</sup>      | Numeric          | Measure: Input/Output Operations per Second (IOPS)    |
```

Reviewer 2: We have a rule to not use acronyms. If we did, it would be `DiskIops` based on our current naming standard.

Reviewer 4: And we could go update this to be one of the supported things? or maybe we don't need to yet because these aren't columns (yet?)

To me, the issue is the term InputOutput as a quantifiable measure means very little ("Hey Azure, what's your DiskInputOutput?")  and the kicker is the description we're clearly describing IOPS. Thanks, agree it needs to be DiskIops according to the way we handle other acronyms. Updated.

Reviewer 2: I get your point. I'm indifferent. InputOutput feels cleaner, but I also get that Iops is more precise. I'm happy to go with whatever the group wants. If we can't close it, then I would just remove it for 1.2 and discuss for 1.3. We'll continue to evolve these keys over time.

Reviewer 4: Feel free to resolve this based on consensus. I saw a couple of people agreeing to Iops below. I wouldn't let the 'no acronym' rule hold back clarity. We've already made exceptions for similar cases

Reviewer 3: Agree with @UdamDewaraja's suggestion that this should be `DiskIops`.  I believe @rileyjenk agrees, based on his comment on [this thread](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/683#discussion_r2017345297).

Reviewer 3: We agreed in Mar 31 Maintainers to sustain this change.

Reviewer 3: @kk09v I am unresolving this comment ("sustain" means "accept"!).  Please either apply the suggested change or provide context for why you do not support it.

Reviewer 3: Upon further review -- looks like @kk09v applied this via a separate commit.  Current version uses `Iops`.  Re-resolving.

| DiskSpace                | Storage capacity available                                      | Numeric          | Measure: Gibibytes (GiB)                              |
| DiskType                 | Kind of disk used                                               | String           | Examples: "SSD", "HDD", "NVMe"                        |
| GpuCount                 | Number of GPUs available                                        | Numeric          | Measure: Quantity of GPUs                             |
| InstanceType             | Common name of the instance including size, shape, series, etc. | String           | Examples: "m5d.2xlarge", "NC24rs_v3", "P50"           |
| MemorySize               | RAM allocated for processing                                    | Numeric          | Measure: Gibibytes (GiB<sup>2</sup>)                  |
| NetworkBandwidth         | Network capacity for data transfer<br><sup>1</sup>              | Numeric          | Measure: Megabits per second (Mbps)                   |

> | NetworkInputOutput       | Network input/output operations per second<br><sup>1</sup>      | Numeric          | Measure: Input/Output Operations per Second (IOPS)    |

Reviewer 4: ```suggestion
| NetworkIOPS       | Network input/output operations per second<br><sup>1</sup>      | Numeric          | Measure: Input/Output Operations per Second (IOPS)    |
```

Reviewer 3: +1 on this change.

Reviewer 3: We agreed in Mar 31 Maintainers to sustain this change.

Reviewer 3: @kk09v I am unresolving this comment ("sustain" means "accept"!).  Please either apply the suggested change or provide context for why you do not support it.

Reviewer 4: "sustain" always confuses me :) 😅

| OperatingSystem          | Operating system family<sup>3</sup>                             | String           | Examples: "Linux", "MacOS", "Windows"                 |
| Redundancy               | Level of redundancy offered by the SKU                          | String           | Examples: "Local", "Zonal", "Global"                  |
| Series                   | Common name for the series and/or generation of the SKU         | String           | Examples: "M5", "Dadv5", "N2D"                        |

> | SizeClass                | Logical grouping of size, which may be used for size fexibility | String           | Examples: "Large", "2xlarge", "Metal"                 |

Reviewer 4: IMO, I think these two are very confusing and are unlike the other properties in this list. These seem like generic 'we can't really describe what all would go in this but here's a key for you to provide whatever you see fit' type of property where as the others seem like they're referring to a specific attribute that isn't so ambiguous. My suggestion is to remove these OR make them very specific like the other things above. E.g., Series could be InstanceFamily which is consistent with the other property above called InstanceType. 

If, instead, this is not an Instance (VM) property (which contradicts the other properties), it feels like we would need a better way to convey what this represents, or providers will put whatever they feel fits here and then this property won't deliver any of the intended value we're hoping for it to deliver.

Reviewer 2: We discussed this at length with @rileyjenk. But if we can't close it, I would vote we remove it and discuss for 1.3.
```suggestion
```

Reviewer 3: I would be disappointed to see `Series` and `SizeClass` get cut.  `InstanceType` alone tells the basic story, but then the practitioner still has to perform provider-specific transformations to get to these smaller chunks of the story.   As a vendor, we found that allowing our customers to build a report that slices cost by `medium`, `large`, `2xlarge`, and so on to be really important, especially if the data can be enriched by observability (e.g., CPU %, MB %, IOPS) to then start unlocking optimizations (e.g., this cost is all associated with `large` instances that should be upsized).  I suspect @kk09v is nodding violently in agreement as he reads this, but I'll let him chime in.

That said, I won't die on this hill if we can't get consensus on this one.

Reviewer 3: We agreed in Mar 31 Maintainers to cut the `Series` and `SizeClass` keys from this PR and consider adding them in a future release, potentially along with more precise guidance on what belongs in such keys.

Author: @shawnalpay I disagree with your summary.
Series is being renamed to InstanceSeries (to align with InstanceType prefix) and clarification of the provider mappings has been added to the supporting content.
[AWS](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/blob/26837645fd5e7163cb482de3fdcdb309a4f2d38f/supporting_content/columns/skupricedetails.md?plain=1#L15)
[GCP](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/blob/26837645fd5e7163cb482de3fdcdb309a4f2d38f/supporting_content/columns/skupricedetails.md?plain=1#L23)
[Azure](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/blob/26837645fd5e7163cb482de3fdcdb309a4f2d38f/supporting_content/columns/skupricedetails.md?plain=1#L38)

I'm working to get alignment with the people who provided feedback above (yourself included) to see if we can remove objections before TF2 on Wednesday.

Author: SizeClass I don't immediately have a path forward for. I haven't found a specific use outside of AWS, so I'm checking with Riley.

Reviewer 3: (Oh... sorry Karl, I must have misremembered the outcome of the conversation.  I'm still nursing a head cold.)

As we agreed in today's call (at least, I'm pretty sure), `Series` means multiple things.  I quote from your supporting content:

> To remove ambiguity, the InstanceSeries Focus-defined property value will refer to the subset of [Azure Virtual Machine naming convention](https://learn.microsoft.com/en-us/azure/virtual-machines/vm-naming-conventions) excluding the "# of vCPUs" and "*Constrained vCPUs" components, as shown in the [Virtual Machine name breakdowns structure](https://learn.microsoft.com/en-us/azure/virtual-machines/sizes/overview#name-structure-breakdown)

This is a stricture on the meaning of the spec that we have placed outside the spec, and no provider is required to adhere to it.  Thus, the practitioner's understanding of this column may be entirely divorced from what a provider puts there (in this case, Azure).  **I think that these keys need more clarity _within the spec_ as to what values the practitioner can expect to find.**  While I am absolutely willing to accept waiting until the next release to install that clarity, I deeply worry about the practitioners' -- and providers' -- understanding of these two keys in the interim.
